File: src/main/java/scampi/cp/core/CPVarBool.java
/*******************************************************************************
 * This program and the accompanying materials
 * are made available under the terms of the GNU Lesser Public License v3
 * which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/lgpl.html
 *  
 * Contributors:
 *      www.n-side.com
 ******************************************************************************/
package scampi.cp.core;


import scampi.cp.constraints.Eq;
import scampi.cp.constraints.Implication;
import scampi.cp.constraints.Or;
import scampi.cp.constraints.Sum;

/**
 * Boolean variable: it is nothing else than a 0-1 integer variable. <br>
 * 1 is used for true, 0 for false.
 * @author Pierre Schaus pschaus@gmail.com
 */
public class CPVarBool extends CPVarInt{

    /**
     *
     * @param n > 0
     * @return an array of n fresh boolean variables
     */
    public static CPVarBool[] getArray(Store s, int n) {
        assert(n > 0);
        CPVarBool [] res = new CPVarBool[n];
        for (int i = 0; i < n; i++) {
            res[i] = new CPVarBool(s);
        }
        return res;
    }

    
    public static CPVarBool[][] getArray(Store s, int n, int m) {
        assert(n > 0);
        assert(m > 0);
        CPVarBool [][] res = new CPVarBool[n][m];
        for (int i = 0; i < n; i++) {
        	for (int j = 0; j < m; j++) {
        		res[i][j] = new CPVarBool(s);
        	}
            
        }
        return res;
    }
    
    
	public CPVarBool(Store s) {
		super(s,0,1);
	}
	
	@Override
	public String toString() {
		if (isBound()) {
			return getValue() == 0 ? "false" : "true";
		} else {
			return "false,true";
		}
	}

    /**
     * @return  a constraint setting the boolean variable to true (1)
     */
    public Constraint constraintTrue() {
        return new Eq(this,1);
    }

    /**
     * @return  a constraint setting the boolean variable to false (0)
     */
    public Constraint constraintFalse() {
        return new Eq(this,0);
    }

    /**
     * @return true if the variable is bound and bound to value 1
     */
    public boolean isTrue() {
        return  isBound() && getValue() == 1;
    }

    /**
     * @return true if the variable is bound and bound to value 0
     */
    public boolean isFalse() {
        return  isBound() && getValue() == 0;
    }
    
    
	/**
	 * Logical or
	 */
	public CPVarBool or(CPVarBool y) {
		CPVarBool b = new CPVarBool(getStore());
		getStore().post(new Or(new CPVarBool[]{this,y},b));
		return b;
	} 
	
	/**
	 * Logical and
	 */
	public CPVarBool and(CPVarBool y) {
		CPVarInt res = this.plus(y);
		return res.isEq(2);
	} 
	
	public CPVarBool implies(CPVarBool y) {
		CPVarBool V = new CPVarBool(getStore());
		getStore().post(new Implication(this, y, V));
		return V;
	}

	//--------------------methods for the scala wrapper--------------------
    
	/**
	 * Scala wrapper: x | y
	 */
	public CPVarBool $bar(CPVarBool y) {
		return this.or(y);
	} 
    
	/**
	 * Scala wrapper: x || y
	 */
	public CPVarBool $bar$bar(CPVarBool y) {
		return this.or(y);
	}  

	/**
	 * Scala wrapper: x & y
	 */
	public CPVarBool $amp(CPVarBool y) {
		return this.and(y);
	}
	
	/**
	 * Scala wrapper: x && y
	 */
	public CPVarBool $amp$amp(CPVarBool y) {
		return this.and(y);
	}
	
	/**
	 * Scala wrapper: x => y
	 */
	public CPVarBool $eq$eq$greater(CPVarBool y) {
		return this.implies(y);
	}





}

File: oscar-cp/src/main/scala/oscar/cp/core/CPVarBool.scala
package oscar.cp.core

import oscar.cp.constraints.Not

/**
 * Boolean variable: it is nothing else than a 0-1 integer variable. <br>
 * 1 is used for true, 0 for false.
 *
 * @author Pierre Schaus pschaus@gmail.com
 * @author Renaud Hartert ren.hartert@gmail.com
 */
class CPVarBool(st: CPStore, name: String) extends CPVarIntImpl(st, 0, 1, name) {

  def this(s: CPStore) = this(s, "")

  def this(s: CPStore, b: Boolean, name: String = "") = {
    this(s, name)
    if (b) assign(1)
    else assign(0)
  }

  override def toString() = {
    if (!isBound) "false, true"
    else if (value == 0) "false"
    else "true"
  }

  /** @return  a constraint setting the boolean variable to true (1) */
  def constraintTrue(): Constraint = new oscar.cp.constraints.Eq(this, 1)

  /** @return  a constraint setting the boolean variable to false (0) */
  def constraintFalse(): Constraint = new oscar.cp.constraints.Eq(this, 0)

  /** @return true if the variable is bound and bound to value 1 */
  def isTrue: Boolean = isBound && value == 1

  /** @return true if the variable is bound and bound to value 0 */
  def isFalse: Boolean = isBound && value == 0

  /** Logical or */
  def or(y: CPVarBool): CPVarBool = {
    val b = new CPVarBool(s)
    s.post(new oscar.cp.constraints.Or(Array(this, y), b))
    b
  }

  /** Logical and */
  def and(y: CPVarBool): CPVarBool = {
    val res = this.plus(y)
    res.isEq(2)
  }

  def not(): CPVarBool = {
    val not = new CPVarBool(s)
    s.post(new oscar.cp.constraints.Not(this, not))
    not
  }

  def implies(y: CPVarBool) = {
    val V = new CPVarBool(s)
    s.post(new oscar.cp.constraints.Implication(this, y, V))
    V
  }

  /**
   * -b
   */
  def unary_!(): CPVarBool = this.not()
  /**
   * x | y
   */
  def |(y: CPVarBool) = this.or(y)
  /**
   * x || y
   */
  def ||(y: CPVarBool) = this.or(y)
  /**
   * x & y
   */
  def &(y: CPVarBool) = this.and(y)
  /**
   * x && y
   */
  def &&(y: CPVarBool) = this.and(y)
  /**
   * x ==> y
   */
  def ==>(y: CPVarBool) = this.implies(y)
  /**
   * x != y
   */
  def !=(y: CPVarBool) = new Not(this, y)
}

object CPVarBool {

  /** Creates a new CP Boolean Variable */
  def apply(name: String)(implicit store: CPStore): CPVarBool = new CPVarBool(store, name)

  /** Creates a new CP Boolean Variable */
  def apply()(implicit store: CPStore): CPVarBool = apply("")(store)

  /** Creates a new CP Boolean Variable assigned to b */
  def apply(b: Boolean, name: String)(implicit store: CPStore): CPVarBool = new CPVarBool(store, b, name)

  /** Creates a new CP Boolean Variable assigned to b */
  def apply(b: Boolean)(implicit store: CPStore): CPVarBool = apply(b, "")(store)

  @deprecated("use apply(b: Boolean)(implicit store: CPStore) instead", "1.0")
  def apply(cp: CPStore, b: Boolean): CPVarBool = new CPVarBool(cp, b)
}  
  
